lib/dialogue-agent/entity-linking/entity_lookup.js

```diff
@@ -34,15 +61,41 @@ async function lookupEntity(dlg, entityType, entityDisplay, idEntities) {
         const candidates = await dlg.manager.platform.getCapability('app-launcher').listApps();
         return getBestEntityMatch(entityDisplay, entityType, candidates);
     }
-
-    const { data:candidates, meta } = await dlg.manager.thingpedia.lookupEntity(entityType, entityDisplay);
-
+    const { data: candidates, meta } = await dlg.manager.thingpedia.lookupEntity(entityType, entityDisplay);
     if (candidates.length === 0) {
-        await dlg.replyInterp(dlg._("Sorry, I cannot find any ${entity_type} matching “${name}”."), {
-            entity_type: meta.name,
-            name: entityDisplay
-        });
-        return null;
+        try {
+            const kind = entityType.split(":")[0];
+            const query = entityType.split(":")[1];
+            const schema = await dlg.manager.schemas.getSchemaAndNames(kind, 'query', query);
+            const filter = new Ast.BooleanExpression.Atom(null, "id", '=~', new Ast.Value.String(entityDisplay));
+            var invocation = (new Ast.Invocation(null, new Ast.Selector.Device(null, kind, null, null), query, [], schema));
+            const invocationTable = new Ast.Table.Invocation(null, invocation, schema);
+            const filteredTable = new Ast.Table.Filter(null, invocationTable, filter, schema);
+            const program = new Ast.Program(null, [], [],
+                [new Ast.Statement.Command(null, filteredTable, [Ast.Action.notifyAction()])], null);
+            const compiler = new AppCompiler(dlg.manager.schemas);
+            const compiled = await compiler.compileProgram(program);
+            const app = await dlg.engine.createApp(program);
+            const mainOutput = new QueueOutputDelegate();
+            const command = new RuleExecutor(dlg.engine, app, compiled.command, mainOutput);
+            command.start();
```

Comment:
Wait! You're executing this twice... If you call createApp, that also executes the app, and then you can just iterate `app.mainOutput` to get the results. See for example: https://github.com/stanford-oval/genie-toolkit/blob/master/lib/dialogue-agent/execution/statement_executor.js#L167

That way you don't have to duplicate so much code, and you don't have to break the abstraction to poke at engine internals like AppCompiler and RuleExecutor.
In fact, I would separate the code that constructs the query program (it can be in this file or it can be somewhere else) from the code executing the query program. And it would be nice to be able to go through the same paths as all other execution of ThingTalk code, because you need to select the device ID and stuff.